core/internal/filetransfer/file_transfer_default.go

```diff
@@ -150,6 +190,112 @@ func (ft *DefaultFileTransfer) Download(task *DefaultDownloadTask) error {
 	return nil
 }
 
+// DownloadMultipart download the file using multiple http range requests.
+// You should always use [Download] instead of this method unless you are testing file write errorh handling.
+func (ft *DefaultFileTransfer) DownloadMultipart(task *DefaultDownloadTask, file WriterAt) error {
+	// Split file into multiple http range requests and distribute to multiple workers
+	requests := SplitHttpRanges(task.Size, ft.multipartDownloadPartSize)
+	numWorkers := min(runtime.NumCPU(), len(requests))
+	workerTasks, err := SplitWorkerTasks(len(requests), numWorkers)
+	if err != nil {
+		return fmt.Errorf("file transfer: download multipart: failed to split worker tasks: %w", err)
+	}
+	// Use channel to stream the chunks from download workers to the single file writer
+	chunkChan := make(chan *FileChunk, numWorkers*10)
+
+	// TODO: We should change the signature to pass the ctx as first argument instead of part of task struct
+	ctx := task.Context
+
+	// We have two error groups. One for download to make sure it close the channel to signal success completion to the file writer.
+	// Another for download and file writer to make sure if one of the has error, the other one would stop.
+	wg, ctx := errgroup.WithContext(ctx)
+	wg.Go(func() error {
+		defer close(chunkChan)
+
+		downloadGroup, ctx := errgroup.WithContext(ctx)
+		for workerID := range numWorkers {
```

Comment:
This looks like a use-case for [errgroup.Group::SetLimit](https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit). Why not start a separate goroutine per range?

Then you don't even need two errgroups. Just start one errgroup with a limit of `numWorkers+1`, start the file-writing goroutine first, then loop over `requests` checking for context cancellation and then calling `Go()`.